File: problem4.php
<?php
//solution: 906609
echo eulerProblem4() . "\n";

function eulerProblem4() {
    $max = 0;
    for($a = 999; $a >= 100; --$a) {
        for($b = 999; $b >= 100; --$b) {
            $ab = $a * $b;
            if(isPalindrome($ab . "") && $ab > $max) {
                $max = $ab;
            }
        }
    }
    return $max;
}

function isPalindrome($string) {
    $low = 0; $high = strlen($string) - 1;
    while($low < $high) {
        if($string{$low} != $string{$high}) {
            return 0;
        }
        $low++; $high--;
    }
    return 1;
}

?>

File: cpp/problem4.cpp
#include <iostream>
using namespace std;
//solution: 906609

bool isPalindrome(char*);

int eulerProblem4() {
    int max = 0;
    char* str = (char*)malloc(sizeof(char)*20); //should check the return value
    for(int i = 100; i < 1000; ++i) {
        for(int j = 100; j < 1000; ++j) {
            int p = i*j;
            sprintf(str, "%d", p);
            if(isPalindrome(str) && p > max) {
                max = p;
            }
        }
    }
    return max;
}

int main() {
    cout << eulerProblem4() << endl;
    return 0;
}

bool isPalindrome(char* str) {
    int len = strlen(str);
    int low = 0; int high = len - 1;
    while(low < high) {
        if(str[low] != str[high]) {
            return false;
        }
        ++low; --high;
    }
    return true;
}
